core/unix/signal.c

```diff
@@ -2293,17 +2293,29 @@ check_signals_pending(dcontext_t *dcontext, thread_sig_info_t *info)
 /* Returns whether to execute the syscall */
 bool
 handle_sigprocmask(dcontext_t *dcontext, int how, kernel_sigset_t *app_set,
-                   kernel_sigset_t *oset, size_t sigsetsize)
+                   kernel_sigset_t *oset, size_t sigsetsize, uint *errno_val)
 {
     thread_sig_info_t *info = (thread_sig_info_t *)dcontext->signal_field;
     int i;
-    kernel_sigset_t safe_set;
+    kernel_sigset_t safe_set, safe_old_set;
+    if ((app_set != NULL && !d_r_safe_read(app_set, sizeof(safe_set), &safe_set)) ||
+        (oset != NULL && !d_r_safe_read(oset, sizeof(safe_old_set), &safe_old_set))) {
```

Comment:
Added check for write privileges for old set, thanks for pointing it out.

Reply:
For the other cases (future write attempts failing etc) I guess there's no way to guarantee success of future writes. Added a comment about that. Also added a TODO to re-use the read sigset instead of making the kernel read the same memory again. Will handle it in a follow up PR so that we can get this one submitted.